File: src/main/java/com/crud_javatokotiln/service/UserService.java
package com.crud_javatokotiln.service;


import com.crud_javatokotiln.dto.UserDto;
import com.crud_javatokotiln.entity.User;
import com.crud_javatokotiln.exception.NotFoundException;
import com.crud_javatokotiln.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class UserService {

    private final UserRepository userRepository;

    public List<User> findAll() {
        return userRepository.findAll();
    }

    public User findById(Long id) {
        return userRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("User not found with id: " + id));
    }

    @Transactional
    public User create(UserDto userDto) {
        User user = new User(userDto.getName(), userDto.getEmail());
        return userRepository.save(user);
    }

    @Transactional
    public User update(Long id, UserDto userDto) {
        User user = findById(id);
        user.setName(userDto.getName());
        user.setEmail(userDto.getEmail());
        return userRepository.save(user);
    }

    @Transactional
    public void delete(Long id) {
        userRepository.deleteById(id);
    }
}

File: src/main/java/com/crud_javatokotiln/service/UserService.kt
package com.crud_javatokotiln.service


import com.crud_javatokotiln.dto.UserDto
import com.crud_javatokotiln.entity.User
import com.crud_javatokotiln.exception.NotFoundException
import com.crud_javatokotiln.repository.UserRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional


@Service
@Transactional(readOnly = true)
open class UserService(
    private val userRepository: UserRepository
) {

    fun findAll(): List<User> = userRepository.findAll().filterNotNull()

    fun findById(id: Long): User =
        userRepository.findById(id)
            .orElseThrow { NotFoundException("User not found with id: $id") }!!

    @Transactional
    open fun create(userDto: UserDto): User {
        val name = userDto.name ?: throw IllegalArgumentException("Name cannot be null")
        val email = userDto.email ?: throw IllegalArgumentException("Email cannot be null")
        val user = User(email, name)
        return userRepository.save(user)
    }

    @Transactional
    open fun update(id: Long, userDto: UserDto): User {
        val user = findById(id)
        user.name = userDto.name ?: user.name
        user.email = userDto.email ?: user.email
        return userRepository.save(user)
    }

    @Transactional
    open fun delete(id: Long) {
        userRepository.deleteById(id)
    }
}
